<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysing Long Read RNA-Seq data with bambu • bambu</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analysing Long Read RNA-Seq data with bambu">
<meta property="og:description" content="bambu">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bambu</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bambu.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/GoekeLab/bambu/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="bambu_files/header-attrs-2.10/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analysing Long Read RNA-Seq data with bambu</h1>
                        <h4 class="author">Ying Chen, Yuk Kei Wan, Jonathan Göke</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/GoekeLab/bambu/blob/master/vignettes/bambu.Rmd"><code>vignettes/bambu.Rmd</code></a></small>
      <div class="hidden name"><code>bambu.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><em><a href="https://github.com/GoekeLab/bambu">Bambu</a></em> is a method for transcript discovery and quantification from long read RNA-Seq data. Bambu uses aligned reads and genome reference annotations as input, and will return abundance estimates for all known transcripts and for newly discovered transcripts. Bambu uses the information from the reference annotations to correct misalignment at splice junctions, then reduces the aligned reads to read equivalent classes, and uses this information to identify novel transcripts across all samples of interest. Reads are then assigned to transcripts, and expression estimates are obtained using an expectation maximisation algorithm. Here, we present an example workflow for analysing Nanopore long read RNA-Sequencing data from two human cancer cell lines from the Singapore Nanopore Expression Project (SG-NEx).</p>
</div>
<div id="quick-start" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick start: Transcript discovery and quantification with bambu</h1>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install bambu from github:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"bambu"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"NanoporeRNASeq"</span><span class="op">)</span></code></pre></div>
</div>
<div id="general-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#general-usage" class="anchor"></a>General Usage</h2>
<p>The default mode to run <em>bambu</em> is using a set of aligned reads (bam files), reference genome annotations (gtf file, TxDb object, or bambuAnnotation object), and reference genome sequence (fasta file or BSgenome). <em>bambu</em> will return a summarizedExperiment object with the genomic coordinates for annotated and new transcripts and transcript expression estimates. We highly recommend to use the same annotations that were used for genome alignment. If you have a gtf file and fasta file you can run bambu with the following options:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(bambu)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>test.bam <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"SGNex_A549_directRNA_replicate5_run1_chr9_1_1000000.bam"</span>,</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">package =</span> <span class="st">"bambu"</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>fa.file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"Homo_sapiens.GRCh38.dna_sm.primary_assembly_chr9_1_1000000.fa"</span>,</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">package =</span> <span class="st">"bambu"</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>gtf.file <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"Homo_sapiens.GRCh38.91_chr9_1_1000000.gtf"</span>, <span class="at">package =</span> <span class="st">"bambu"</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>bambuAnnotations <span class="ot">&lt;-</span> <span class="fu">prepareAnnotations</span>(gtf.file)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">bambu</span>(<span class="at">reads =</span> test.bam, <span class="at">annotations =</span> bambuAnnotations, <span class="at">genome =</span> fa.file)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="do">##&gt; </span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>  <span class="er">|</span>                                                                      <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================================================|</span> <span class="dv">100</span>%</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="do">##&gt; </span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="do">##&gt; </span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>  <span class="er">|</span>                                                                      <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================================================|</span> <span class="dv">100</span>%</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="do">##&gt; </span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a><span class="do">##&gt; </span></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>  <span class="er">|</span>                                                                      <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================================================|</span> <span class="dv">100</span>%</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a><span class="do">##&gt; </span></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a><span class="do">##&gt; </span></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>  <span class="er">|</span>                                                                      <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================================================|</span> <span class="dv">100</span>%</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a><span class="do">##&gt; </span></span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a><span class="do">##&gt; </span></span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>  <span class="er">|</span>                                                                      <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================================================|</span> <span class="dv">100</span>%</span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a><span class="do">##&gt; </span></span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a><span class="do">##&gt; </span></span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb2-41"><a href="#cb2-41" aria-hidden="true" tabindex="-1"></a>  <span class="er">|</span>                                                                      <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb2-42"><a href="#cb2-42" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb2-43"><a href="#cb2-43" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================================================|</span> <span class="dv">100</span>%</span></code></pre></div>
<p><em>bambu</em> returns a SummarizedExperiment object which can be accessed as follows:</p>
<ul>
<li>assays(se) returns the transcript abundance estimates as counts or CPM<br>
</li>
<li>rowRanges(se) returns a GRangesList with all annotated and newly discovered transcripts<br>
</li>
<li>rowData(se) returns additional information about each transcript such as the gene name and the class of newly discovered transcript</li>
</ul>
</div>
</div>
<div id="a-complete-workflow-to-identify-and-quantify-transcript-expression-from" class="section level1">
<h1 class="hasAnchor">
<a href="#a-complete-workflow-to-identify-and-quantify-transcript-expression-from" class="anchor"></a>A complete workflow to identify and quantify transcript expression from</h1>
<p>Nanopore RNA-Seq data {#complete-workflow} To demonstrate the usage of Bambu, we used long-read RNA-Seq data generated using Oxford Nanopore Sequencing from the NanoporeRNASeq package, which consists of 6 samples from two human cell lines (K562 and MCF7) that were generated by the <a href="https://github.com/GoekeLab/sg-nex-data">SG-NEx project</a>. Each of these cell lines has three replicates, with 1 direct RNA sequencing run and 2 cDNA sequencing runs. Reads are aligned to chromosome 22 (Grch38) and stored as bam files. In this workflow, we will demonstrate how to apply <em>bambu</em> to these bam files to identify novel transcripts and estimates transcript expression, visualize the results, and identify differentially expressed genes and transcripts.</p>
<div id="input-data" class="section level2">
<h2 class="hasAnchor">
<a href="#input-data" class="anchor"></a>Input data</h2>
<div id="bam-files" class="section level3">
<h3 class="hasAnchor">
<a href="#bam-files" class="anchor"></a>Aligned reads (bam files)</h3>
<p><em>bambu</em> takes genomic alignments saved in bam files. Here we use bam-files from the <em>NanoporeRNASeq</em> package, which contains reads aligned to the first half of the human chromosome 22 using <em>minimap2</em>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/GoekeLab/bambu">bambu</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/GoekeLab/NanoporeRNASeq">NanoporeRNASeq</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"SGNexSamples"</span><span class="op">)</span>
<span class="va">SGNexSamples</span>
<span class="co">##&gt; DataFrame with 6 rows and 6 columns</span>
<span class="co">##&gt;                sample_id    Platform    cellLine    protocol cancer_type</span>
<span class="co">##&gt;              &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="co">##&gt; 1 K562_directcDNA_repl..      MinION        K562  directcDNA   Leukocyte</span>
<span class="co">##&gt; 2 K562_directcDNA_repl..     GridION        K562  directcDNA   Leukocyte</span>
<span class="co">##&gt; 3 K562_directRNA_repli..     GridION        K562   directRNA   Leukocyte</span>
<span class="co">##&gt; 4 MCF7_directcDNA_repl..      MinION        MCF7  directcDNA      Breast</span>
<span class="co">##&gt; 5 MCF7_directcDNA_repl..     GridION        MCF7  directcDNA      Breast</span>
<span class="co">##&gt; 6 MCF7_directRNA_repli..     GridION        MCF7   directRNA      Breast</span>
<span class="co">##&gt;                fileNames</span>
<span class="co">##&gt;              &lt;character&gt;</span>
<span class="co">##&gt; 1 NanoporeRNASeq/versi..</span>
<span class="co">##&gt; 2 NanoporeRNASeq/versi..</span>
<span class="co">##&gt; 3 NanoporeRNASeq/versi..</span>
<span class="co">##&gt; 4 NanoporeRNASeq/versi..</span>
<span class="co">##&gt; 5 NanoporeRNASeq/versi..</span>
<span class="co">##&gt; 6 NanoporeRNASeq/versi..</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span>
<span class="va">NanoporeData</span> <span class="op">&lt;-</span> <span class="fu">query</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NanoporeRNA"</span>, <span class="st">"GRCh38"</span>, <span class="st">"BAM"</span><span class="op">)</span><span class="op">)</span>
<span class="va">bamFiles</span> <span class="op">&lt;-</span> <span class="fu">Rsamtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html">BamFileList</a></span><span class="op">(</span><span class="va">NanoporeData</span><span class="op">[[</span><span class="st">"EH3808"</span><span class="op">]</span><span class="op">]</span>, <span class="va">NanoporeData</span><span class="op">[[</span><span class="st">"EH3809"</span><span class="op">]</span><span class="op">]</span>,
    <span class="va">NanoporeData</span><span class="op">[[</span><span class="st">"EH3810"</span><span class="op">]</span><span class="op">]</span>, <span class="va">NanoporeData</span><span class="op">[[</span><span class="st">"EH3811"</span><span class="op">]</span><span class="op">]</span>, <span class="va">NanoporeData</span><span class="op">[[</span><span class="st">"EH3812"</span><span class="op">]</span><span class="op">]</span>,
    <span class="va">NanoporeData</span><span class="op">[[</span><span class="st">"EH3813"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
</div>
<div id="genome-sequences" class="section level3">
<h3 class="hasAnchor">
<a href="#genome-sequences" class="anchor"></a>Genome sequence (fasta file/ BSGenome object)</h3>
<p><em>bambu</em> additionally requires a genome sequence, which is used to correct splicing junctions in read alignments. Ideally, we recommend to use the same genome seqeunce file that was used for alignment to be used for bambu.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get path to fasta file</span>
<span class="va">genomeSequenceData</span> <span class="op">&lt;-</span> <span class="fu">query</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NanoporeRNA"</span>, <span class="st">"GRCh38"</span>, <span class="st">"FASTA"</span><span class="op">)</span><span class="op">)</span>
<span class="va">genomeSequence</span> <span class="op">&lt;-</span> <span class="va">genomeSequenceData</span><span class="op">[[</span><span class="st">"EH7260"</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p>As an option, users can also choose to use a BSgenome object:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.NCBI.GRCh38</span><span class="op">)</span>
<span class="va">genomeSequenceBsgenome</span> <span class="op">&lt;-</span> <span class="va">BSgenome.Hsapiens.NCBI.GRCh38</span></code></pre></div>
</div>
<div id="genome-annotations-bambu-annotations-object-gtf-file-txdb-object" class="section level3">
<h3 class="hasAnchor">
<a href="#genome-annotations-bambu-annotations-object-gtf-file-txdb-object" class="anchor"></a>Genome annotations (bambu annotations object/ gtf file / TxDb object)</h3>
<p>{#annotations}<br><em>bambu</em> also requires a reference transcript annotations object, which is used to correct read alignments, to identify for transcripts and genes (and the type for novel transcripts), and for quantification. The annotation object can be created from a gtf file:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gtf.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Homo_sapiens.GRCh38.91_chr9_1_1000000.gtf"</span>, package <span class="op">=</span> <span class="st">"bambu"</span><span class="op">)</span>
<span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareAnnotations.html">prepareAnnotations</a></span><span class="op">(</span><span class="va">gtf.file</span><span class="op">)</span></code></pre></div>
<p>The annotation object can also be created from a TxDb object:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Homo_sapiens.GRCh38.91_chr9_1_1000000.gtf"</span>, package <span class="op">=</span> <span class="st">"bambu"</span><span class="op">)</span>
<span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareAnnotations.html">prepareAnnotations</a></span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span></code></pre></div>
<p>The annotation object can be stored and used again for re-running bambu. Here we will used the annotation object from the <em>NanoporeRNASeq</em> package that wasis prepared from a gtf file using the  function in by function in <em>bambu</em>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"HsChr22BambuAnnotation"</span><span class="op">)</span>
<span class="va">HsChr22BambuAnnotation</span>
<span class="co">##&gt; GRangesList object of length 1500:</span>
<span class="co">##&gt; $ENST00000043402</span>
<span class="co">##&gt; GRanges object with 2 ranges and 2 metadata columns:</span>
<span class="co">##&gt;       seqnames            ranges strand | exon_rank exon_endRank</span>
<span class="co">##&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;    &lt;integer&gt;</span>
<span class="co">##&gt;   [1]       22 20241415-20243110      - |         2            1</span>
<span class="co">##&gt;   [2]       22 20268071-20268531      - |         1            2</span>
<span class="co">##&gt;   -------</span>
<span class="co">##&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="co">##&gt; </span>
<span class="co">##&gt; $ENST00000086933</span>
<span class="co">##&gt; GRanges object with 3 ranges and 2 metadata columns:</span>
<span class="co">##&gt;       seqnames            ranges strand | exon_rank exon_endRank</span>
<span class="co">##&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;    &lt;integer&gt;</span>
<span class="co">##&gt;   [1]       22 19148576-19149095      - |         3            1</span>
<span class="co">##&gt;   [2]       22 19149663-19149916      - |         2            2</span>
<span class="co">##&gt;   [3]       22 19150025-19150283      - |         1            3</span>
<span class="co">##&gt;   -------</span>
<span class="co">##&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="co">##&gt; </span>
<span class="co">##&gt; $ENST00000155674</span>
<span class="co">##&gt; GRanges object with 8 ranges and 2 metadata columns:</span>
<span class="co">##&gt;       seqnames            ranges strand | exon_rank exon_endRank</span>
<span class="co">##&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;    &lt;integer&gt;</span>
<span class="co">##&gt;   [1]       22 17137511-17138357      - |         8            1</span>
<span class="co">##&gt;   [2]       22 17138550-17138738      - |         7            2</span>
<span class="co">##&gt;   [3]       22 17141059-17141233      - |         6            3</span>
<span class="co">##&gt;   [4]       22 17143098-17143131      - |         5            4</span>
<span class="co">##&gt;   [5]       22 17145024-17145117      - |         4            5</span>
<span class="co">##&gt;   [6]       22 17148448-17148560      - |         3            6</span>
<span class="co">##&gt;   [7]       22 17149542-17149745      - |         2            7</span>
<span class="co">##&gt;   [8]       22 17165209-17165287      - |         1            8</span>
<span class="co">##&gt;   -------</span>
<span class="co">##&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="co">##&gt; </span>
<span class="co">##&gt; ...</span>
<span class="co">##&gt; &lt;1497 more elements&gt;</span></code></pre></div>
</div>
</div>
<div id="transcript-discovery-quantification" class="section level2">
<h2 class="hasAnchor">
<a href="#transcript-discovery-quantification" class="anchor"></a>Transcript discovery and quantification</h2>
<div id="run-bambu" class="section level3">
<h3 class="hasAnchor">
<a href="#run-bambu" class="anchor"></a>Running bambu</h3>
<p>Next we apply <em>bambu</em> on the input data (bam files, annotations, genomeSequence). Bambu will perform isoform discovery to extend the provided annotation, and then quantify the transcript expression from these extended annotation using an Expectation-Maximisation algorithm. Here we will use 1 core, which can be changed to process multiple files in parallel.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bambu.html">bambu</a></span><span class="op">(</span>reads <span class="op">=</span> <span class="va">bamFiles</span>, annotations <span class="op">=</span> <span class="va">HsChr22BambuAnnotation</span>, genome <span class="op">=</span> <span class="va">genomeSequence</span>,
    ncore <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">se</span>
<span class="co">##&gt; class: RangedSummarizedExperiment </span>
<span class="co">##&gt; dim: 1511 6 </span>
<span class="co">##&gt; metadata(0):</span>
<span class="co">##&gt; assays(6): counts CPM ... uniqueCounts theta</span>
<span class="co">##&gt; rownames(1511): tx.1 tx.2 ... ENST00000641933 ENST00000641967</span>
<span class="co">##&gt; rowData names(6): TXNAME GENEID ... readCount txNDR</span>
<span class="co">##&gt; colnames(6): a59435bcc4d4_3844 a5946402f226_3846 ... a594dde6847_3852</span>
<span class="co">##&gt;   a5947ea78412_3854</span>
<span class="co">##&gt; colData names(3): name d_rate nGeneFordRate</span></code></pre></div>
<p>For the downstream analysis, we will add the condition of interest to the  object that describes the samples. Here we are interested in a comparison of the 2 cell lines:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">SGNexSamples</span><span class="op">$</span><span class="va">cellLine</span><span class="op">)</span></code></pre></div>
<p>Optionally, users can choose to apply <em>bambu</em> to do quantification only (without isoform discovery)</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seUnextended</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bambu.html">bambu</a></span><span class="op">(</span>reads <span class="op">=</span> <span class="va">bamFiles</span>, annotations <span class="op">=</span> <span class="va">HsChr22BambuAnnotation</span>, genome <span class="op">=</span> <span class="va">genomeSequence</span>,
    discovery <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seUnextended</span>
<span class="co">##&gt; class: RangedSummarizedExperiment </span>
<span class="co">##&gt; dim: 1500 6 </span>
<span class="co">##&gt; metadata(0):</span>
<span class="co">##&gt; assays(6): counts CPM ... uniqueCounts theta</span>
<span class="co">##&gt; rownames(1500): ENST00000043402 ENST00000086933 ... ENST00000641933</span>
<span class="co">##&gt;   ENST00000641967</span>
<span class="co">##&gt; rowData names(3): TXNAME GENEID eqClass</span>
<span class="co">##&gt; colnames(6): a59435bcc4d4_3844 a5946402f226_3846 ... a594dde6847_3852</span>
<span class="co">##&gt;   a5947ea78412_3854</span>
<span class="co">##&gt; colData names(3): name d_rate nGeneFordRate</span></code></pre></div>
</div>
<div id="visualise-results" class="section level3">
<h3 class="hasAnchor">
<a href="#visualise-results" class="anchor"></a>Visualise results</h3>
<p><em>bambu</em> provides functions to visualise and explore the results. When multiple samples are used, we can visualise the correlation and clustering of all samples with a heatmap:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="fu"><a href="../reference/plotBambu.html">plotBambu</a></span><span class="op">(</span><span class="va">se</span>, type <span class="op">=</span> <span class="st">"heatmap"</span><span class="op">)</span></code></pre></div>
<p><img src="bambu_files/figure-html/unnamed-chunk-15-1.png" width="768"></p>
<p>Additionally, we can also visualise the correlation with a 2-dimmensional PCA plot.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotBambu.html">plotBambu</a></span><span class="op">(</span><span class="va">se</span>, type <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></code></pre></div>
<p><img src="bambu_files/figure-html/unnamed-chunk-16-1.png" width="768"></p>
<p>In addition to visualising the correlation between samples, <em>bambu</em> also provide a function to visualise the extended annotation and expression estimation for individual genes. Here we look at gene ENSG00000099968 and visualise the transcript coordinates for annotated and novel isoforms and expression levels for these isoforms across all samples.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotBambu.html">plotBambu</a></span><span class="op">(</span><span class="va">se</span>, type <span class="op">=</span> <span class="st">"annotation"</span>, gene_id <span class="op">=</span> <span class="st">"ENSG00000099968"</span><span class="op">)</span></code></pre></div>
<p><img src="bambu_files/figure-html/unnamed-chunk-17-1.png" width="768"></p>
<pre><code>##&gt; [[1]]
##&gt; TableGrob (3 x 1) "arrange": 3 grobs
##&gt;   z     cells    name                grob
##&gt; 1 1 (2-2,1-1) arrange      gtable[layout]
##&gt; 2 2 (3-3,1-1) arrange      gtable[layout]
##&gt; 3 3 (1-1,1-1) arrange text[GRID.text.250]</code></pre>
</div>
<div id="obtain-gene-expression-estimates-from-transcript-expression" class="section level3">
<h3 class="hasAnchor">
<a href="#obtain-gene-expression-estimates-from-transcript-expression" class="anchor"></a>Obtain gene expression estimates from transcript expression</h3>
<p>{#gene-expression} Gene expression can be calculated from the transcript expression estimates returned by <em>bambu</em> using the  function. Looking at the output, we can see there are novel genes identified as well</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seGene</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transcriptToGeneExpression.html">transcriptToGeneExpression</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>
<span class="va">seGene</span>
<span class="co">##&gt; class: RangedSummarizedExperiment </span>
<span class="co">##&gt; dim: 574 6 </span>
<span class="co">##&gt; metadata(0):</span>
<span class="co">##&gt; assays(2): counts CPM</span>
<span class="co">##&gt; rownames(574): ENSG00000015475 ENSG00000040608 ... gene.43 gene.45</span>
<span class="co">##&gt; rowData names(2): GENEID newGeneClass</span>
<span class="co">##&gt; colnames(6): a59435bcc4d4_3844 a5946402f226_3846 ... a594dde6847_3852</span>
<span class="co">##&gt;   a5947ea78412_3854</span>
<span class="co">##&gt; colData names(4): name d_rate nGeneFordRate condition</span></code></pre></div>
<p>We can again use the  function to visualise the gene expression data across the 6 samples with a heatmap or PCA plot. As expected, samples from the same cell line showed higher correlation than across the cell lines.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">seGene</span><span class="op">)</span><span class="op">$</span><span class="va">groupVar</span> <span class="op">&lt;-</span> <span class="va">SGNexSamples</span><span class="op">$</span><span class="va">cellLine</span>
<span class="fu"><a href="../reference/plotBambu.html">plotBambu</a></span><span class="op">(</span><span class="va">seGene</span>, type <span class="op">=</span> <span class="st">"heatmap"</span><span class="op">)</span></code></pre></div>
<p><img src="bambu_files/figure-html/unnamed-chunk-19-1.png" width="768"></p>
</div>
<div id="save-data" class="section level3">
<h3 class="hasAnchor">
<a href="#save-data" class="anchor"></a>Save data (gtf/text)</h3>
<p><em>bambu</em> includes a function to write the extended annotations, the transcript and the gene expression estimates that include any newly discovered genes and transcripts to text files.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">save.dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="../reference/writeBambuOutput.html">writeBambuOutput</a></span><span class="op">(</span><span class="va">se</span>, path <span class="op">=</span> <span class="va">save.dir</span>, prefix <span class="op">=</span> <span class="st">"NanoporeRNASeq_"</span><span class="op">)</span></code></pre></div>
<p><em>bambu</em> also includes a function that only exports the extended annotations to gtf file:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">save.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".gtf"</span><span class="op">)</span>
<span class="fu"><a href="../reference/writeToGTF.html">writeToGTF</a></span><span class="op">(</span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">se</span><span class="op">)</span>, file <span class="op">=</span> <span class="va">save.file</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="DESeq2" class="section level2">
<h2 class="hasAnchor">
<a href="#DESeq2" class="anchor"></a>Identifying differentially expressed genes</h2>
<p>One of the most common tasks when analysing RNA-Seq data is the analysis of differential gene expression across a condition of intertest. Here we use <em>DESeq2</em> to find the differentially expressed genes between MCF7 and K562 cell lines. Similar to using results from <em>Salmon</em>, estimates from <em>bambu</em> will first be rounded.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2">DESeq2</a></span><span class="op">)</span>
<span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html">DESeqDataSetFromMatrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu">assays</span><span class="op">(</span><span class="va">seGene</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>, colData <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span>,
    design <span class="op">=</span> <span class="op">~</span><span class="va">condition</span><span class="op">)</span>
<span class="va">dds.deseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html">DESeq</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span>
<span class="va">deGeneRes</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">results</a></span><span class="op">(</span><span class="va">dds.deseq</span>, independentFiltering <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">deGeneRes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">deGeneRes</span><span class="op">$</span><span class="va">padj</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span>
<span class="co">##&gt; log2 fold change (MLE): condition MCF7 vs K562 </span>
<span class="co">##&gt; Wald test p-value: condition MCF7 vs K562 </span>
<span class="co">##&gt; DataFrame with 6 rows and 6 columns</span>
<span class="co">##&gt;                  baseMean log2FoldChange     lfcSE      stat</span>
<span class="co">##&gt;                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="co">##&gt; ENSG00000185686  500.6470       -7.15159  0.500364 -14.29278</span>
<span class="co">##&gt; ENSG00000283633   95.7518       -9.10519  1.246517  -7.30451</span>
<span class="co">##&gt; ENSG00000197077   26.9189        9.17563  1.320390   6.94918</span>
<span class="co">##&gt; ENSG00000240972 2443.3934        2.47082  0.357369   6.91392</span>
<span class="co">##&gt; ENSG00000099977  235.8601        1.92963  0.306609   6.29347</span>
<span class="co">##&gt; ENSG00000169635   43.5245       -3.36897  0.569649  -5.91411</span>
<span class="co">##&gt;                                                                pvalue</span>
<span class="co">##&gt;                                                             &lt;numeric&gt;</span>
<span class="co">##&gt; ENSG00000185686 0.000000000000000000000000000000000000000000000242728</span>
<span class="co">##&gt; ENSG00000283633 0.000000000000278275679947613751071839495840056657549</span>
<span class="co">##&gt; ENSG00000197077 0.000000000003674139901044937399445365779334784031053</span>
<span class="co">##&gt; ENSG00000240972 0.000000000004714473225561034087277656270998130953048</span>
<span class="co">##&gt; ENSG00000099977 0.000000000310450546572304490812751211792114465481962</span>
<span class="co">##&gt; ENSG00000169635 0.000000003336692881462347780346996492929594058107767</span>
<span class="co">##&gt;                                                                padj</span>
<span class="co">##&gt;                                                           &lt;numeric&gt;</span>
<span class="co">##&gt; ENSG00000185686 0.0000000000000000000000000000000000000000000631093</span>
<span class="co">##&gt; ENSG00000283633 0.0000000000361758383931897894568746600884185651500</span>
<span class="co">##&gt; ENSG00000197077 0.0000000003064407596614672382912675321117289967132</span>
<span class="co">##&gt; ENSG00000240972 0.0000000003064407596614672382912675321117289967132</span>
<span class="co">##&gt; ENSG00000099977 0.0000000161434284217598349698291349809883832300272</span>
<span class="co">##&gt; ENSG00000169635 0.0000001445900248633684085023933191607725667893192</span></code></pre></div>
<p>A quick summary of differentially expressed genes</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">deGeneRes</span><span class="op">)</span>
<span class="co">##&gt; </span>
<span class="co">##&gt; out of 260 with nonzero total read count</span>
<span class="co">##&gt; adjusted p-value &lt; 0.1</span>
<span class="co">##&gt; LFC &gt; 0 (up)       : 19, 7.3%</span>
<span class="co">##&gt; LFC &lt; 0 (down)     : 22, 8.5%</span>
<span class="co">##&gt; outliers [1]       : 0, 0%</span>
<span class="co">##&gt; low counts [2]     : 0, 0%</span>
<span class="co">##&gt; (mean count &lt; 0)</span>
<span class="co">##&gt; [1] see 'cooksCutoff' argument of ?results</span>
<span class="co">##&gt; [2] see 'independentFiltering' argument of ?results</span></code></pre></div>
<p>We can also visualise the MA-plot for differentially used isoforms using . However, visualizing the MA-plots using the original log-fold change results will be affected by the noise associated with log2 fold changes from low count genes without requiring arbitrary filtering thresholds. As recommended in the [<em>DESeq2</em>tutorial] (<a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/" class="uri">http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/</a> DESeq2.html#alternative-shrinkage-estimators). we applied the same shrinkage to effect sizes to improve the visualization.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">apeglm</span><span class="op">)</span>
<span class="va">resLFC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/lfcShrink.html">lfcShrink</a></span><span class="op">(</span><span class="va">dds.deseq</span>, coef <span class="op">=</span> <span class="st">"condition_MCF7_vs_K562"</span>, type <span class="op">=</span> <span class="st">"apeglm"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/plotMA.html">plotMA</a></span><span class="op">(</span><span class="va">resLFC</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="bambu_files/figure-html/unnamed-chunk-24-1.png" width="768"></p>
</div>
<div id="DEXSeq" class="section level2">
<h2 class="hasAnchor">
<a href="#DEXSeq" class="anchor"></a>Identifying differential transcript usage</h2>
<p>We used <em>DEXSeq</em> to detect alternative used isoforms.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">DEXSeq</span><span class="op">)</span>
<span class="va">dxd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DEXSeq/man/DEXSeqDataSet.html">DEXSeqDataSet</a></span><span class="op">(</span>countData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu">assays</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>, sampleData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span>,
    design <span class="op">=</span> <span class="op">~</span><span class="va">sample</span> <span class="op">+</span> <span class="va">exon</span> <span class="op">+</span> <span class="va">condition</span><span class="op">:</span><span class="va">exon</span>, featureID <span class="op">=</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">TXNAME</span>, groupID <span class="op">=</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">GENEID</span><span class="op">)</span>
<span class="va">dxr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DEXSeq/man/DEXSeq.html">DEXSeq</a></span><span class="op">(</span><span class="va">dxd</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">dxr</span><span class="op">)</span>
<span class="co">##&gt; </span>
<span class="co">##&gt; LRT p-value: full vs reduced</span>
<span class="co">##&gt; </span>
<span class="co">##&gt; DataFrame with 6 rows and 12 columns</span>
<span class="co">##&gt;                              groupID   featureID exonBaseMean dispersion</span>
<span class="co">##&gt;                          &lt;character&gt; &lt;character&gt;    &lt;numeric&gt;  &lt;numeric&gt;</span>
<span class="co">##&gt; gene.30:tx.1                 gene.30        tx.1     9.059656         NA</span>
<span class="co">##&gt; gene.35:tx.2                 gene.35        tx.2     1.046280         NA</span>
<span class="co">##&gt; gene.12:tx.3                 gene.12        tx.3     1.054917         NA</span>
<span class="co">##&gt; ENSG00000184702:tx.4 ENSG00000184702        tx.4    32.603977   0.904926</span>
<span class="co">##&gt; ENSG00000183506:tx.5 ENSG00000183506        tx.5    14.804874   1.399612</span>
<span class="co">##&gt; gene.45:tx.6                 gene.45        tx.6     0.767463         NA</span>
<span class="co">##&gt;                           stat    pvalue      padj      K562      MCF7</span>
<span class="co">##&gt;                      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="co">##&gt; gene.30:tx.1                NA        NA        NA        NA        NA</span>
<span class="co">##&gt; gene.35:tx.2                NA        NA        NA        NA        NA</span>
<span class="co">##&gt; gene.12:tx.3                NA        NA        NA        NA        NA</span>
<span class="co">##&gt; ENSG00000184702:tx.4   1.86587  0.171949  0.591465   1.31079   1.64896</span>
<span class="co">##&gt; ENSG00000183506:tx.5   1.32103  0.250408  0.631578   1.06310   1.20895</span>
<span class="co">##&gt; gene.45:tx.6                NA        NA        NA        NA        NA</span>
<span class="co">##&gt;                      log2fold_MCF7_K562   genomicData     countData</span>
<span class="co">##&gt;                               &lt;numeric&gt; &lt;GRangesList&gt;      &lt;matrix&gt;</span>
<span class="co">##&gt; gene.30:tx.1                         NA                 36:48:5:...</span>
<span class="co">##&gt; gene.35:tx.2                         NA                   9:5:0:...</span>
<span class="co">##&gt; gene.12:tx.3                         NA                   8:3:2:...</span>
<span class="co">##&gt; ENSG00000184702:tx.4           1.162939               24:132:88:...</span>
<span class="co">##&gt; ENSG00000183506:tx.5           0.522927                43:61:24:...</span>
<span class="co">##&gt; gene.45:tx.6                         NA                   5:5:0:...</span></code></pre></div>
<p>We can visualize the MA-plot</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/DEXSeq/man/plotMA.html">plotMA</a></span><span class="op">(</span><span class="va">dxr</span>, cex <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></code></pre></div>
<p><img src="bambu_files/figure-html/unnamed-chunk-26-1.png" width="768"></p>
</div>
</div>
<div id="large-sample-num" class="section level1">
<h1 class="hasAnchor">
<a href="#large-sample-num" class="anchor"></a>Running bambu with large number of samples</h1>
<p>For larger sample numbers we recommend to write the processed data to a file. This can be done by providing the readClass.outputDir:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bambu.html">bambu</a></span><span class="op">(</span>reads <span class="op">=</span> <span class="va">bamFiles</span>, rcOutDir <span class="op">=</span> <span class="st">"./bambu/"</span>, annotations <span class="op">=</span> <span class="va">annotaiton</span>, genome <span class="op">=</span> <span class="va">genomeSequence</span><span class="op">)</span></code></pre></div>
</div>
<div id="bambu-parameters" class="section level1">
<h1 class="hasAnchor">
<a href="#bambu-parameters" class="anchor"></a>Bambu parameters</h1>
<div id="advanced-options" class="section level2">
<h2 class="hasAnchor">
<a href="#advanced-options" class="anchor"></a>Advanced Options</h2>
<p>For transcript discovery we recommend to adjust the parameters according to the number of replicates and the sequencing throughput. The most relevant parameters are explained here. You can use any combination of these parameters. ### More stringent filtering thresholds imposed on potential novel transcripts</p>
<ul>
<li>Keep novel transcripts with min 5 read count in at least 1 sample:</li>
</ul>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/bambu.html">bambu</a></span><span class="op">(</span><span class="va">reads</span>, <span class="va">annotations</span>, <span class="va">genome</span>, opt.discovery <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>min.readCount <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<ul>
<li>Keep novel transcripts with min 5 samples having at least 2 counts:</li>
</ul>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/bambu.html">bambu</a></span><span class="op">(</span><span class="va">reads</span>, <span class="va">annotations</span>, <span class="va">genome</span>, opt.discovery <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>min.sampleNumber <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<ul>
<li>Filter out transcripts with relative abundance within gene lower than 10%:</li>
</ul>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/bambu.html">bambu</a></span><span class="op">(</span><span class="va">reads</span>, <span class="va">annotations</span>, <span class="va">genome</span>, opt.discovery <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>min.readFractionByGene <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div id="quantification-without-bias-correction" class="section level3">
<h3 class="hasAnchor">
<a href="#quantification-without-bias-correction" class="anchor"></a>Quantification without bias correction</h3>
<p>The default estimation automatically does bias correction for expression estimates. However, you can choose to perform the quantification without bias correction.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/bambu.html">bambu</a></span><span class="op">(</span><span class="va">reads</span>, <span class="va">annotations</span>, <span class="va">genome</span>, opt.em <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>bias <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="parallel-computation" class="section level3">
<h3 class="hasAnchor">
<a href="#parallel-computation" class="anchor"></a>Parallel computation</h3>
<p>bambu allows parallel computation.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/bambu.html">bambu</a></span><span class="op">(</span><span class="va">reads</span>, <span class="va">annotations</span>, <span class="va">genome</span>, ncore <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></code></pre></div>
<p>See <em><a href="https://github.com/GoekeLab/bambu/blob/master/docs/bambu_0.1.0.pdf">manual</a></em> for details to customize other conditions.</p>
</div>
</div>
</div>
<div id="get-help" class="section level1">
<h1 class="hasAnchor">
<a href="#get-help" class="anchor"></a>Getting help</h1>
<p>Questions and issues can be raised at the Bioconductor support site (once bambu is available through bioconductor): <a href="https://support.bioconductor.org" class="uri">https://support.bioconductor.org</a>. Please tag your your posts with bambu.</p>
<p>Alternatively, issues can be raised at the bambu Github repository:<a href="https://github.com/GoekeLab/bambu" class="uri">https://github.com/GoekeLab/bambu</a>.</p>
</div>
<div id="cite-bambu" class="section level1">
<h1 class="hasAnchor">
<a href="#cite-bambu" class="anchor"></a>Citing bambu</h1>
<p>A manuscript describing bambu is currently in preparation. If you use bambu for your research, please cite using the following doi: 10.5281/zenodo.3900025.</p>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Information</h1>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">##&gt; R version 4.1.0 (2021-05-18)</span>
<span class="co">##&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">##&gt; Running under: Ubuntu 20.04.3 LTS</span>
<span class="co">##&gt; </span>
<span class="co">##&gt; Matrix products: default</span>
<span class="co">##&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span>
<span class="co">##&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span>
<span class="co">##&gt; </span>
<span class="co">##&gt; locale:</span>
<span class="co">##&gt;  [1] LC_CTYPE=en_SG.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##&gt;  [3] LC_TIME=en_SG.UTF-8        LC_COLLATE=en_SG.UTF-8    </span>
<span class="co">##&gt;  [5] LC_MONETARY=en_SG.UTF-8    LC_MESSAGES=en_SG.UTF-8   </span>
<span class="co">##&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">##&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">##&gt; [11] LC_MEASUREMENT=en_SG.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">##&gt; </span>
<span class="co">##&gt; attached base packages:</span>
<span class="co">##&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">##&gt; [8] base     </span>
<span class="co">##&gt; </span>
<span class="co">##&gt; other attached packages:</span>
<span class="co">##&gt;  [1] DEXSeq_1.39.0                         </span>
<span class="co">##&gt;  [2] RColorBrewer_1.1-2                    </span>
<span class="co">##&gt;  [3] AnnotationDbi_1.55.1                  </span>
<span class="co">##&gt;  [4] BiocParallel_1.27.4                   </span>
<span class="co">##&gt;  [5] apeglm_1.15.0                         </span>
<span class="co">##&gt;  [6] DESeq2_1.33.4                         </span>
<span class="co">##&gt;  [7] ggplot2_3.3.5                         </span>
<span class="co">##&gt;  [8] BSgenome.Hsapiens.NCBI.GRCh38_1.3.1000</span>
<span class="co">##&gt;  [9] BSgenome_1.61.0                       </span>
<span class="co">##&gt; [10] rtracklayer_1.53.1                    </span>
<span class="co">##&gt; [11] Rsamtools_2.9.1                       </span>
<span class="co">##&gt; [12] Biostrings_2.61.2                     </span>
<span class="co">##&gt; [13] XVector_0.33.0                        </span>
<span class="co">##&gt; [14] NanoporeRNASeq_1.3.3                  </span>
<span class="co">##&gt; [15] ExperimentHub_2.1.4                   </span>
<span class="co">##&gt; [16] AnnotationHub_3.1.5                   </span>
<span class="co">##&gt; [17] BiocFileCache_2.1.1                   </span>
<span class="co">##&gt; [18] dbplyr_2.1.1                          </span>
<span class="co">##&gt; [19] bambu_1.99.0                          </span>
<span class="co">##&gt; [20] SummarizedExperiment_1.23.4           </span>
<span class="co">##&gt; [21] Biobase_2.53.0                        </span>
<span class="co">##&gt; [22] GenomicRanges_1.45.0                  </span>
<span class="co">##&gt; [23] GenomeInfoDb_1.29.4                   </span>
<span class="co">##&gt; [24] IRanges_2.27.2                        </span>
<span class="co">##&gt; [25] S4Vectors_0.31.2                      </span>
<span class="co">##&gt; [26] BiocGenerics_0.39.2                   </span>
<span class="co">##&gt; [27] MatrixGenerics_1.5.3                  </span>
<span class="co">##&gt; [28] matrixStats_0.60.1                    </span>
<span class="co">##&gt; </span>
<span class="co">##&gt; loaded via a namespace (and not attached):</span>
<span class="co">##&gt;   [1] utf8_1.2.2                    tidyselect_1.1.1             </span>
<span class="co">##&gt;   [3] RSQLite_2.2.8                 htmlwidgets_1.5.3            </span>
<span class="co">##&gt;   [5] grid_4.1.0                    munsell_0.5.0                </span>
<span class="co">##&gt;   [7] codetools_0.2-18              ragg_1.1.3                   </span>
<span class="co">##&gt;   [9] statmod_1.4.36                xgboost_1.4.1.1              </span>
<span class="co">##&gt;  [11] withr_2.4.2                   colorspace_2.0-2             </span>
<span class="co">##&gt;  [13] filelock_1.0.2                OrganismDbi_1.35.0           </span>
<span class="co">##&gt;  [15] highr_0.9                     knitr_1.33                   </span>
<span class="co">##&gt;  [17] rstudioapi_0.13               labeling_0.4.2               </span>
<span class="co">##&gt;  [19] bbmle_1.0.24                  GenomeInfoDbData_1.2.6       </span>
<span class="co">##&gt;  [21] hwriter_1.3.2                 bit64_4.0.5                  </span>
<span class="co">##&gt;  [23] farver_2.1.0                  rprojroot_2.0.2              </span>
<span class="co">##&gt;  [25] coda_0.19-4                   vctrs_0.3.8                  </span>
<span class="co">##&gt;  [27] generics_0.1.0                xfun_0.25                    </span>
<span class="co">##&gt;  [29] biovizBase_1.41.0             R6_2.5.1                     </span>
<span class="co">##&gt;  [31] doParallel_1.0.16             clue_0.3-59                  </span>
<span class="co">##&gt;  [33] locfit_1.5-9.4                AnnotationFilter_1.17.1      </span>
<span class="co">##&gt;  [35] bitops_1.0-7                  cachem_1.0.6                 </span>
<span class="co">##&gt;  [37] reshape_0.8.8                 DelayedArray_0.19.1          </span>
<span class="co">##&gt;  [39] assertthat_0.2.1              promises_1.2.0.1             </span>
<span class="co">##&gt;  [41] BiocIO_1.3.0                  scales_1.1.1                 </span>
<span class="co">##&gt;  [43] nnet_7.3-16                   gtable_0.3.0                 </span>
<span class="co">##&gt;  [45] ggbio_1.41.0                  ensembldb_2.17.4             </span>
<span class="co">##&gt;  [47] rlang_0.4.11                  genefilter_1.75.1            </span>
<span class="co">##&gt;  [49] systemfonts_1.0.3             GlobalOptions_0.1.2          </span>
<span class="co">##&gt;  [51] splines_4.1.0                 lazyeval_0.2.2               </span>
<span class="co">##&gt;  [53] dichromat_2.0-0               checkmate_2.0.0              </span>
<span class="co">##&gt;  [55] BiocManager_1.30.16           yaml_2.2.1                   </span>
<span class="co">##&gt;  [57] reshape2_1.4.4                GenomicFeatures_1.45.1       </span>
<span class="co">##&gt;  [59] backports_1.2.1               httpuv_1.6.2                 </span>
<span class="co">##&gt;  [61] Hmisc_4.5-0                   RBGL_1.69.0                  </span>
<span class="co">##&gt;  [63] tools_4.1.0                   ellipsis_0.3.2               </span>
<span class="co">##&gt;  [65] jquerylib_0.1.4               Rcpp_1.0.7                   </span>
<span class="co">##&gt;  [67] plyr_1.8.6                    base64enc_0.1-3              </span>
<span class="co">##&gt;  [69] progress_1.2.2                zlibbioc_1.39.0              </span>
<span class="co">##&gt;  [71] purrr_0.3.4                   RCurl_1.98-1.4               </span>
<span class="co">##&gt;  [73] prettyunits_1.1.1             rpart_4.1-15                 </span>
<span class="co">##&gt;  [75] GetoptLong_1.0.5              cluster_2.1.2                </span>
<span class="co">##&gt;  [77] fs_1.5.0                      magrittr_2.0.1               </span>
<span class="co">##&gt;  [79] data.table_1.14.0             circlize_0.4.13              </span>
<span class="co">##&gt;  [81] mvtnorm_1.1-2                 ProtGenerics_1.25.1          </span>
<span class="co">##&gt;  [83] hms_1.1.0                     mime_0.11                    </span>
<span class="co">##&gt;  [85] evaluate_0.14                 xtable_1.8-4                 </span>
<span class="co">##&gt;  [87] XML_3.99-0.7                  emdbook_1.3.12               </span>
<span class="co">##&gt;  [89] jpeg_0.1-9                    gridExtra_2.3                </span>
<span class="co">##&gt;  [91] shape_1.4.6                   bdsmatrix_1.3-4              </span>
<span class="co">##&gt;  [93] compiler_4.1.0                biomaRt_2.49.4               </span>
<span class="co">##&gt;  [95] tibble_3.1.4                  crayon_1.4.1                 </span>
<span class="co">##&gt;  [97] htmltools_0.5.2               later_1.3.0                  </span>
<span class="co">##&gt;  [99] Formula_1.2-4                 tidyr_1.1.3                  </span>
<span class="co">##&gt; [101] geneplotter_1.71.0            DBI_1.1.1                    </span>
<span class="co">##&gt; [103] formatR_1.11                  ComplexHeatmap_2.9.3         </span>
<span class="co">##&gt; [105] MASS_7.3-54                   rappdirs_0.3.3               </span>
<span class="co">##&gt; [107] Matrix_1.3-4                  parallel_4.1.0               </span>
<span class="co">##&gt; [109] pkgconfig_2.0.3               pkgdown_1.6.1                </span>
<span class="co">##&gt; [111] GenomicAlignments_1.29.0      numDeriv_2016.8-1.1          </span>
<span class="co">##&gt; [113] foreign_0.8-81                xml2_1.3.2                   </span>
<span class="co">##&gt; [115] foreach_1.5.1                 annotate_1.71.0              </span>
<span class="co">##&gt; [117] bslib_0.2.5.1                 stringr_1.4.0                </span>
<span class="co">##&gt; [119] VariantAnnotation_1.39.0      digest_0.6.27                </span>
<span class="co">##&gt; [121] graph_1.71.2                  rmarkdown_2.10               </span>
<span class="co">##&gt; [123] htmlTable_2.2.1               restfulr_0.0.13              </span>
<span class="co">##&gt; [125] curl_4.3.2                    shiny_1.6.0                  </span>
<span class="co">##&gt; [127] rjson_0.2.20                  lifecycle_1.0.0              </span>
<span class="co">##&gt; [129] jsonlite_1.7.2                desc_1.3.0                   </span>
<span class="co">##&gt; [131] fansi_0.5.0                   pillar_1.6.2                 </span>
<span class="co">##&gt; [133] lattice_0.20-44               GGally_2.1.2                 </span>
<span class="co">##&gt; [135] KEGGREST_1.33.0               fastmap_1.1.0                </span>
<span class="co">##&gt; [137] httr_1.4.2                    survival_3.2-13              </span>
<span class="co">##&gt; [139] interactiveDisplayBase_1.31.2 glue_1.4.2                   </span>
<span class="co">##&gt; [141] png_0.1-7                     iterators_1.0.13             </span>
<span class="co">##&gt; [143] BiocVersion_3.14.0            bit_4.0.4                    </span>
<span class="co">##&gt; [145] stringi_1.7.4                 sass_0.4.0                   </span>
<span class="co">##&gt; [147] blob_1.2.2                    textshaping_0.3.6            </span>
<span class="co">##&gt; [149] latticeExtra_0.6-29           memoise_2.0.0                </span>
<span class="co">##&gt; [151] dplyr_1.0.7</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ying Chen, Andre Sim, Yuk Kei Wan, Jonathan Goeke.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
