<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>long read isoform reconstruction and quantification — bambu • bambu</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="long read isoform reconstruction and quantification — bambu" />
<meta property="og:description" content="This function takes bam file of genomic alignments and performs
isoform recontruction and gene and transcript expression quantification.
It also allows saving of read class files of alignments, extending provided
annotations, and quantification based on extended annotations. When multiple
samples are provided, extended annotations will be combined across samples to
allow comparison." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bambu</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bambu.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/GoekeLab/bambu/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>long read isoform reconstruction and quantification</h1>
    <small class="dont-index">Source: <a href='https://github.com/GoekeLab/bambu/blob/master/R/bambu.R'><code>R/bambu.R</code></a></small>
    <div class="hidden name"><code>bambu.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function takes bam file of genomic alignments and performs
isoform recontruction and gene and transcript expression quantification.
It also allows saving of read class files of alignments, extending provided
annotations, and quantification based on extended annotations. When multiple
samples are provided, extended annotations will be combined across samples to
allow comparison.</p>
    </div>

    <pre class="usage"><span class='fu'>bambu</span><span class='op'>(</span>
    reads <span class='op'>=</span> <span class='cn'>NULL</span>,
    rcFile <span class='op'>=</span> <span class='cn'>NULL</span>,
    rcOutDir <span class='op'>=</span> <span class='cn'>NULL</span>,
    annotations <span class='op'>=</span> <span class='cn'>NULL</span>,
    genome <span class='op'>=</span> <span class='cn'>NULL</span>,
    stranded <span class='op'>=</span> <span class='cn'>FALSE</span>,
    ncore <span class='op'>=</span> <span class='fl'>1</span>,
    NDR <span class='op'>=</span> <span class='fl'>0.1</span>,
    yieldSize <span class='op'>=</span> <span class='cn'>NULL</span>,
    opt.discovery <span class='op'>=</span> <span class='cn'>NULL</span>,
    opt.em <span class='op'>=</span> <span class='cn'>NULL</span>,
    discovery <span class='op'>=</span> <span class='cn'>TRUE</span>,
    quant <span class='op'>=</span> <span class='cn'>TRUE</span>,
    verbose <span class='op'>=</span> <span class='cn'>FALSE</span>,
    lowMemory <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>reads</th>
      <td><p>A string or a vector of strings specifying the paths of bam
files for genomic alignments, or a <code>BamFile</code> object or a
<code>BamFileList</code>  object (see <code>Rsamtools</code>).</p></td>
    </tr>
    <tr>
      <th>rcFile</th>
      <td><p>A string or a vector of strings specifying the read
class files that are saved during previous run of <code>bambu</code>.</p></td>
    </tr>
    <tr>
      <th>rcOutDir</th>
      <td><p>A string variable specifying the path to where
read class files will be saved.</p></td>
    </tr>
    <tr>
      <th>annotations</th>
      <td><p>A <code>TxDb</code> object or A GRangesList object
obtained by <code><a href='prepareAnnotations.html'>prepareAnnotations</a></code>.</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>A fasta file or a BSGenome object.</p></td>
    </tr>
    <tr>
      <th>stranded</th>
      <td><p>A boolean for strandedness, defaults to FALSE.</p></td>
    </tr>
    <tr>
      <th>ncore</th>
      <td><p>specifying number of cores used when parallel processing 
is used, defaults to 1.</p></td>
    </tr>
    <tr>
        <th>NDR</th>
        <td><p>specifying the approximated precision for included novel transcripts, 
defaults to 0.1 (10% precision).</p></td>
      </tr>
    <tr>
      <th>yieldSize</th>
      <td><p>see <code>Rsamtools</code>.</p></td>
    </tr>
    <tr>
      <th>opt.discovery</th>
      <td><p>A list of controlling parameters for isoform
reconstruction process:</p><dl>
    <dt>prefix</dt><dd><p>specifying prefix for new gene Ids (genePrefix.number),
    defaults to empty</p></dd>
    <dt>remove.subsetTx</dt><dd><p>indicating whether filter to remove read classes
    which are a subset of known transcripts(), defaults to TRUE</p></dd>
    <dt>min.readCount</dt><dd><p>specifying minimun read count to consider a read
    class valid in a sample, defaults to 2</p></dd>
    <dt>min.readFractionByGene</dt><dd><p>specifying minimum relative read count per
    gene, highly expressed genes will have many high read count low relative
    abundance transcripts that can be filtered, defaults to 0.05</p></dd>
    <dt>min.sampleNumber</dt><dd><p>specifying minimum sample number with minimum read
    count, defaults to 1</p></dd>
    <dt>min.exonDistance</dt><dd><p>specifying minum distance to known transcript 
    to be considered valid as new, defaults to 35bp</p></dd>
    <dt>min.exonOverlap</dt><dd><p>specifying minimum number of bases shared with
    annotation to be assigned to the same gene id, defaults to 10bp</p></dd>
    <dt>min.primarySecondaryDist</dt><dd><p>specifying the minimum number of distance 
    threshold, defaults to 5bp</p></dd>
    <dt>min.primarySecondaryDistStartEnd1</dt><dd><p>specifying the minimum number 
    of distance threshold, used for extending annotation, defaults to 5bp</p></dd>
    <dt>min.primarySecondaryDistStartEnd2</dt><dd><p>specifying the minimum number 
    of distance threshold, used for estimating distance to annotation, 
    defaults to 5bp</p></dd>
    <dt>min.txScore.multiExon</dt><dd><p>specifying the minimum transcript level 
    threshold for multi-exon transcripts during sample combining, 
    defaults to 0</p></dd>
    <dt>min.txScore.singleExon</dt><dd><p>specifying the minimum transcript level 
    threshold for single-exon transcripts during sample combining, defaults 
    to 1</p></dd>
    <dt>NDR</dt><dd><p>specifying the maximum NDR rate to novel transcript
    output from detected read classes, defaults to 0.1</p></dd>

</dl></td>
    </tr>
    <tr>
      <th>opt.em</th>
      <td><p>A list of controlling parameters for quantification
algorithm estimation process:</p><dl>
    <dt>maxiter</dt><dd><p>specifying maximum number of run iterations,
    defaults to 10000</p></dd>
    <dt>degradationBias</dt><dd><p>correcting for degradation bias, defaults to TRUE</p></dd>
    <dt>conv</dt><dd><p>specifying the covergence threshold control, 
    defaults to 0.0001</p></dd>
    <dt>minvalue</dt><dd><p>specifying the minvalue for convergence consideration, 
    defaults to 0.00000001</p></dd>

</dl></td>
    </tr>
    <tr>
      <th>discovery</th>
      <td><p>A logical variable indicating whether annotations
are to be extended for quantification.</p></td>
    </tr>
    <tr>
      <th>quant</th>
      <td><p>A logical variable indicating whether quantification will 
be performed</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>A logical variable indicating whether processing messages will
be printed.</p></td>
    </tr>
    <tr>
      <th>lowMemory</th>
      <td><p>Read classes will be processed by chromosomes when lowMemory 
is specified. This option provides an efficient way to process big samples.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>bambu</code> will output different results depending on whether
<em>quant</em> mode is on. By default, <em>quant</em> is set to TRUE, so 
<code>bambu</code> will generate a <em>SummarizedExperiment</em> object that contains
the transcript expression estimates. Transcript expression estimates can be 
accessed by <em>counts()</em>, including the following variables</p><dl>
    <dt>counts</dt><dd><p>expression estimates</p></dd>
    <dt>CPM</dt><dd><p>sequencing depth normalized estimates</p></dd>
    <dt>fullLengthCounts</dt><dd><p>estimates of read counts mapped as full length 
    reads for each transcript</p></dd>
    <dt>partialLengthCounts</dt><dd><p>estimates of read counts mapped as partial 
    length reads for each transcript</p></dd>
    <dt>uniqueCounts</dt><dd><p>counts of reads that are uniquely mapped to each 
    transcript</p></dd>
    <dt>theta</dt><dd><p>raw estimates</p></dd>

</dl><p>Output annotations that are usually the annotations with/without novel 
transcripts/genes added, depending on whether <em>discovery</em> mode is on
can be accessed by <em>rowRanges()</em>
Transcript to gene map can be accessed by <em>rowData()</em>, with 
<em>eqClass</em> that defining equivalent class for each transcript</p>
<p>In the case when <em>quant</em> is set to FALSE, i.e., only transcript 
discovery is performed, <code>bambu</code> will report the <em>grangeslist</em> of 
the extended annotations.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Main function</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## =====================</span>
<span class='va'>test.bam</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>,
    <span class='st'>"SGNex_A549_directRNA_replicate5_run1_chr9_1_1000000.bam"</span>,
    package <span class='op'>=</span> <span class='st'>"bambu"</span><span class='op'>)</span>
<span class='va'>fa.file</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, 
    <span class='st'>"Homo_sapiens.GRCh38.dna_sm.primary_assembly_chr9_1_1000000.fa"</span>, 
    package <span class='op'>=</span> <span class='st'>"bambu"</span><span class='op'>)</span>
<span class='va'>gr</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, 
    <span class='st'>"annotationGranges_txdbGrch38_91_chr9_1_1000000.rds"</span>,
    package <span class='op'>=</span> <span class='st'>"bambu"</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>se</span> <span class='op'>&lt;-</span> <span class='fu'>bambu</span><span class='op'>(</span>reads <span class='op'>=</span> <span class='va'>test.bam</span>, annotations <span class='op'>=</span> <span class='va'>gr</span>, 
    genome <span class='op'>=</span> <span class='va'>fa.file</span>, discovery <span class='op'>=</span> <span class='cn'>TRUE</span>, quant <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Start generating read class files</span></div><div class='output co'>#&gt; 
  |                                                                            
  |                                                                      |   0%</div><div class='output co'>#&gt; <span class='message'>not all chromosomes present in reference annotations,</span>
#&gt; <span class='message'>            annotations might be incomplete. Please compare objects</span>
#&gt; <span class='message'>            on the same reference</span></div><div class='output co'>#&gt; <span class='message'>not all chromosomes from reads present in reference genome </span>
#&gt; <span class='message'>            sequence, reads without reference chromosome sequence are dropped</span></div><div class='output co'>#&gt; <span class='message'>Junction correction with not enough data,</span>
#&gt; <span class='message'>            precalculated model is used</span></div><div class='output co'>#&gt; <span class='message'>Not enough data points</span></div><div class='output co'>#&gt; <span class='message'>Not enough TRUE/FALSE labels</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Transcript model not trained. Using pre-trained models</span></div><div class='output co'>#&gt; 
  |                                                                            
  |======================================================================| 100%
#&gt; </div><div class='output co'>#&gt; <span class='message'>Finished generating read classes from genomic alignments.</span></div><div class='output co'>#&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |======================================================================| 100%
#&gt; 
#&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |======================================================================| 100%
#&gt; 
#&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |======================================================================| 100%
#&gt; 
#&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |======================================================================| 100%
#&gt; </div><div class='output co'>#&gt; <span class='warning'>Warning: Less than 50 TRUE or FALSE read classes for precision stabilization. </span>
#&gt; <span class='warning'>          Filtering by prediction score instead</span></div><div class='output co'>#&gt; <span class='message'>Finished extending annotations.</span></div><div class='output co'>#&gt; <span class='message'>Start isoform quantification</span></div><div class='output co'>#&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |======================================================================| 100%
#&gt; </div><div class='output co'>#&gt; <span class='message'>Finished isoform quantification.</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Ying Chen, Andre Sim, Yuk Kei Wan, Jonathan Goeke.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


