<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>prepare reference annotations for long read RNA-Seq analysis with Bambu — prepareAnnotations • bambu</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="prepare reference annotations for long read RNA-Seq analysis with Bambu — prepareAnnotations" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bambu</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.1.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bambu.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/GoekeLab/bambu/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>prepare reference annotations for long read RNA-Seq analysis with Bambu</h1>
    <small class="dont-index">Source: <a href='https://github.com/GoekeLab/bambu/blob/master/R/prepareAnnotations.R'><code>R/prepareAnnotations.R</code></a></small>
    <div class="hidden name"><code>prepareAnnotations.Rd</code></div>
    </div>

    <div class="ref-description">
    </div>

    <pre class="usage"><span class='fu'>prepareAnnotations</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>A path to gtf file or a <code>TxDb</code> object.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <code>GRangesList</code> object with additional details for each exon and transcript that are required by Bambu. Exons are ranked by the <code>exon_rank</code> column, corresponding to the rank in direction of transcription (from first to last exon). In addition to exon rank, gene id, transcript id, and the minimum transcript equivalent class is stored as well (a transcript equivalence class of a transcript x is the collection of transcripts where their exon junctions contain, in a continuous way, the exon junctions of the transcript x). The object is designed to be used by Bambu, and the direct access of the metadata is not recommended.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function creates a reference annotation object which is used for transcript discovery and quantification in Bambu. prepareAnnotations can use a path to a gtf file or a TxDB object as input, and returns a annotation object that stores additional information about transcripts which is used in Bambu. For each transcript, exons are ranked from first to last exon in direction of transcription.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='bambu.html'>bambu()</a></code> for transcript discovery and quantification from long read RNA-Seq.</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='r-in'><span class='va'>gtf.file</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>,</span>
<span class='r-in'>    <span class='st'>"Homo_sapiens.GRCh38.91_chr9_1_1000000.gtf"</span>,</span>
<span class='r-in'>    package <span class='op'>=</span> <span class='st'>"bambu"</span></span>
<span class='r-in'><span class='op'>)</span></span>
<span class='r-in'><span class='va'>annotations</span> <span class='op'>&lt;-</span> <span class='fu'>prepareAnnotations</span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>gtf.file</span><span class='op'>)</span></span>
<span class='r-in'></span>
<span class='r-in'><span class='co'># run bambu</span></span>
<span class='r-in'><span class='va'>test.bam</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>,</span>
<span class='r-in'>    <span class='st'>"SGNex_A549_directRNA_replicate5_run1_chr9_1_1000000.bam"</span>, </span>
<span class='r-in'>    package <span class='op'>=</span> <span class='st'>"bambu"</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>fa.file</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, </span>
<span class='r-in'>    <span class='st'>"Homo_sapiens.GRCh38.dna_sm.primary_assembly_chr9_1_1000000.fa"</span>, </span>
<span class='r-in'>    package <span class='op'>=</span> <span class='st'>"bambu"</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>se</span> <span class='op'>&lt;-</span> <span class='fu'><a href='bambu.html'>bambu</a></span><span class='op'>(</span>reads <span class='op'>=</span> <span class='va'>test.bam</span>, annotations <span class='op'>=</span> <span class='va'>annotations</span>, </span>
<span class='r-in'>    genome <span class='op'>=</span> <span class='va'>fa.file</span>, discovery <span class='op'>=</span> <span class='cn'>TRUE</span>, quant <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> 'getOption("repos")' replaces Bioconductor standard repositories, see</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> '?repositories' for details</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> replacement repositories:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     CRAN: https://cloud.r-project.org</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Start generating read class files</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> iteration: </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> not all chromosomes present in reference annotations,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>             annotations might be incomplete. Please compare objects</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>             on the same reference</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> not all chromosomes from reads present in reference genome </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>             sequence, reads without reference chromosome sequence are dropped</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Junction correction with not enough data,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>             precalculated model is used</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Not enough data points</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Not enough TRUE/FALSE labels</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Transcript model not trained. Using pre-trained models</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> 1</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Finished generating read classes from genomic alignments.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> iteration: </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> 1</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> iteration: </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> 1</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> iteration: </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> 1</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> iteration: </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> 1</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Less than 50 TRUE or FALSE read classes for precision stabilization. </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>           Filtering by prediction score instead</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> -- Predicting annotation completeness to determine NDR threshold --</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Calculated NDR: 0.005</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> all detect novel transcripts are already present in the annotations, try a higher NDR</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> GRangesList object of length 105:</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> $ENST00000190165</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> GRanges object with 2 ranges and 2 metadata columns:</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>       seqnames        ranges strand | exon_rank exon_endRank</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;    &lt;integer&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   [1]        9 976964-977455      + |         1            2</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   [2]        9 990041-991731      + |         2            1</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   -------</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> $ENST00000305248</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> GRanges object with 2 ranges and 2 metadata columns:</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>       seqnames      ranges strand | exon_rank exon_endRank</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;    &lt;integer&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   [1]        9 34965-35264      - |         2            1</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   [2]        9 35504-35871      - |         1            2</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   -------</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> $ENST00000314367</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> GRanges object with 16 ranges and 2 metadata columns:</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>        seqnames        ranges strand | exon_rank exon_endRank</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>           &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;    &lt;integer&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>    [1]        9 121060-121573      - |        16            1</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>    [2]        9 121961-122090      - |        15            2</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>    [3]        9 123217-123282      - |        14            3</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>    [4]        9 123386-123454      - |        13            4</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>    [5]        9 134979-135030      - |        12            5</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>    ...      ...           ...    ... .       ...          ...</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   [12]        9 172081-172172      - |         5           12</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   [13]        9 173270-173366      - |         4           13</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   [14]        9 175698-175784      - |         3           14</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   [15]        9 177723-177820      - |         2           15</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   [16]        9 178816-179047      - |         1           16</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   -------</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ...</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> &lt;102 more elements&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> DataFrame with 105 rows and 10 columns</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                          TXNAME          GENEID      txid</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                     &lt;character&gt;     &lt;character&gt; &lt;logical&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000190165 ENST00000190165 ENSG00000064218        NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000305248 ENST00000305248 ENSG00000218839        NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000314367 ENST00000314367 ENSG00000172785        NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000354485 ENST00000354485 ENSG00000107104        NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000356521 ENST00000356521 ENSG00000172785        NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ...                         ...             ...       ...</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000620292 ENST00000620292 ENSG00000172785        NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000620326 ENST00000620326 ENSG00000277631        NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000621255 ENST00000621255 ENSG00000277631        NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000622412 ENST00000622412 ENSG00000277631        NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000628764 ENST00000628764 ENSG00000227155        NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                                eqClass eqClassById readCount  newTxClass</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                            &lt;character&gt;   &lt;logical&gt; &lt;integer&gt; &lt;character&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000190165        ENST00000190165          NA        NA  annotation</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000305248 ENST00000305248.ENST..          NA        NA  annotation</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000314367        ENST00000314367          NA        NA  annotation</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000354485 ENST00000354485.ENST..          NA        NA  annotation</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000356521 ENST00000356521.ENST..          NA        22  annotation</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ...                                ...         ...       ...         ...</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000620292 ENST00000498044.ENST..          NA        NA  annotation</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000620326        ENST00000620326          NA        NA  annotation</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000621255        ENST00000621255          NA        NA  annotation</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000622412        ENST00000622412          NA        NA  annotation</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000628764        ENST00000628764          NA        NA  annotation</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                     txNDR relReadCount relSubsetCount</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                 &lt;numeric&gt;    &lt;numeric&gt;      &lt;numeric&gt;</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000190165        NA           NA             NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000305248        NA           NA             NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000314367        NA           NA             NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000354485        NA           NA             NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000356521 0.0654418     0.468085              1</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ...                   ...          ...            ...</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000620292        NA           NA             NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000620326        NA           NA             NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000621255        NA           NA             NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000622412        NA           NA             NA</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> ENST00000628764        NA           NA             NA</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Finished extending annotations.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Start isoform quantification</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> iteration: </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> 1</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Finished isoform quantification.</span>
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Ying Chen, Andre Sim, Yuk Kei Wan, Jonathan Goeke.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


